internal/controller/workload_controller.go

```diff
@@ -90,11 +95,191 @@ func (aw *AppWrapper) PodSets() []kueue.PodSet {
 
 func (aw *AppWrapper) RunWithPodSetsInfo(podSetsInfo []podset.PodSetInfo) error {
 	aw.Spec.Suspend = false
-	return nil // TODO
+
+	// Update aw.Spec.Components to inject our labels and Kueue's PodSetInfo into every nested PodTemplateSpec
+	podSetsInfoIndex := 0
+	for componentIndex := range aw.Spec.Components {
+		component := &aw.Spec.Components[componentIndex]
+		objChanged := false
+		obj := &unstructured.Unstructured{}
+		if _, _, err := unstructured.UnstructuredJSONScheme.Decode(component.Template.Raw, nil, obj); err != nil {
+			return err
+		}
+
+		for _, podSet := range component.PodSets {
+			podSetsInfoIndex += 1
+			if podSetsInfoIndex <= len(podSetsInfo) {
+				toInject := podSetsInfo[podSetsInfoIndex-1]
+				parts := strings.Split(podSet.Path, ".")
+				p := obj.UnstructuredContent()
+				var ok bool
+				for i := 1; i < len(parts); i++ {
+					p, ok = p[parts[i]].(map[string]interface{})
+					if !ok {
+						return fmt.Errorf("path element %v not found (segment %v of %v)", parts[i], i, len(parts))
+					}
+				}
+				objChanged = true // Even if currentInfo is empty, we will still add appWrapperLabel to p.metadata.labels
```

Comment:
is this a tautology then? or equivalent to `podSetsInfoIndex > 0` in the end?

Reply:
if there are multiple components, it is possible a particular component does not have any pod specs.

Reply:
Ok I misunderstood the scope of `objChanged`. Thanks!